### DISCUSSION TITLE
"Public API", or prefered way, to call Ruby from Java

### DISCUSSION THREAD

**Discussion by mauricioszabo** [jruby/jruby#6873] [Q&A] [answered]
So, I started a project as a joke: https://github.com/mauricioszabo/clojuby. But then, a real necessity emerged from that code: to somehow "bridge" functions from Java world to JRuby.

At the beginning, I was using `evalScriptlet` to run string fragments, but now I have to map some objects from Java to JRuby. For example, supposing a have a `java.util.List`, I can convert it to JRuby with `RubyArray.newArray` but:

1. Is this the right way? I'm asking because maybe there are faster alternatives, for example, `Ruby.fastNewSymbol`. Also, I'm not sure if I _should be using_ the `fastNewSymbol` either...
2. How do I wrap blocks? This is the most problematic for me right now, because the code that I have to wrap have to use `self`. To have an example that's almost what I want to do, suppose I have this code:

```ruby
"some string".instance_eval do
  self.upcase # <- This code I want to implement in Java
end
```

So, on Java side, I want to somehow generate a `Block`, capture `self` (I know that `instance_eval` can receive a parameter that equates to `self` but the code I'm wrapping does not have this) or any other local var that may have been in scope (for example, on a Sinatra handler, I `params` also enters in the rule) and then convert `self` or the var back to a Java object and do my code in Java side (the conversion I think I already figured out). 

So... how can I do this? Is there a "public API" to make this possible? I'm asking because all the `Block` and `Arity` etc doesn't seem that it's supposed to be used as-is... also, I wasn't able to find documentation for these methods.

**Comment by mohits** [comment, marked as answer]
Hi @mauricioszabo - does this page in the wiki help you?
https://github.com/jruby/jruby/wiki/RedBridge

The pages are a bit old but if you run into issues, please ask here or in the mailing list and you might get some more answers!

**Reply by headius** [reply]
The RedBridge API is probably the best option. The only thing more direct would be calling straight into JRuby internals, which may not be desirable since those APIs can change over time.
